keras_nlp/layers/random_deletion.py

```diff
@@ -0,0 +1,137 @@
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import random
+from typing import Any
+from typing import Dict
+
+import tensorflow as tf
+from tensorflow import keras
+
+
+class RandomDeletion(keras.layers.Layer):
+    """Augments input by randomly deleting words
+
+    Args:
+        probability: probability of a word being chosen for deletion
+        max_deletions: The maximum number of words to replace
+
+    Examples:
+
+    Basic usage.
+    >>> augmenter = keras_nlp.layers.RandomDeletion(
+    ...     probability = 1,
+    ...     max_deletions = 1,
+    ... )
+    >>> augmenter(["dog dog dog dog dog"])
+    <tf.Tensor: shape=(1,), dtype=string, numpy=array([b'dog dog dog dog'],
+    dtype=object)>
+    """
+
+    def __init__(self, probability, max_deletions, **kwargs) -> None:
+        # Check dtype and provide a default.
+        if "dtype" not in kwargs or kwargs["dtype"] is None:
+            kwargs["dtype"] = tf.int32
+        else:
+            dtype = tf.dtypes.as_dtype(kwargs["dtype"])
+            if not dtype.is_integer and dtype != tf.string:
+                raise ValueError(
+                    "Output dtype must be an integer type of a string. "
+                    f"Received: dtype={dtype}"
+                )
+
+        super().__init__(**kwargs)
+        self.probability = probability
+        self.max_deletions = max_deletions
+
+    def call(self, inputs):
+        """Augments input by randomly deleting words
+
+        Args:
+            inputs: A tensor or nested tensor of strings to augment.
+
+        Returns:
+            A tensor or nested tensor of augmented strings.
+        """
+        # If input is a simple String convert it into a list
+        isString = False
+        if isinstance(inputs, str):
+            inputs = [inputs]
+            isString = True
+        # If input is not a tensor convert it into a tensor
+        if not isinstance(inputs, (tf.Tensor, tf.RaggedTensor)):
+            inputs = tf.convert_to_tensor(inputs)
+            inputs = tf.cast(inputs, tf.string)
+
+        def _map_fn(inputs):
+            scalar_input = inputs.shape.rank == 0
+            if scalar_input:
+                inputs = tf.expand_dims(inputs, 0)
+            ragged_words = tf.strings.split(inputs)
+            # Get the row splits for the ragged tensor
+            row_splits = ragged_words.row_splits.numpy()
+            mask = (
+                tf.random.uniform(ragged_words.flat_values.shape)
+                > self.probability
+            )
+            # Iterate to check for any cases where deletions exceed the maximum
+            for i in range(len(row_splits) - 1):
```

Comment:
I think we can clean this whole for loop up, though will need to think about how exactly a bit more.

One place for inspiration is probably the op code for tf text's [RandomItemSelector](https://github.com/tensorflow/text/blob/v2.9.0/tensorflow_text/python/ops/item_selector_ops.py#L225). Which is also selecting a number of items based on a probability with a max cap.

I'm a little skeptical that this would function trace, you are doing a lot of looping and calling `.numpy()`, the later definitely won't work in a compiled context.

Reply:
Reading about RandomItemSelector, I think briefly what it does is to:
1. Calculate how many to select, let's call it N.
2. shuffle the list/array/tensor's index array.
3. Pick the first N elements from index array, then use `tf.gather` to get the actual selected elements.

Reply:
Thanks for sharing this @mattdangerw and for the concise summary @chenmoneygithub this seems like a much more smoother way to do it, will incorporate this idea in my code. Just to clarify do I need to cite this code file in the code as (inspired by) or something of that sort?